File: apps/ios/Sources/Core/Media/MediaContract.swift
import Foundation

// Parent wraps the existing original ConversationScope/account permit. Its check covers
// credential epoch, cache generation, membershipScope and authorizationRevision, never reconstructed IDs.
protocol MediaScope: Sendable {
    var presentationID: String { get }
    var roomID: String? { get }
    func check() throws
}
enum MediaError: Error, Equatable { case invalid, expired, unavailable, processing; case response(Int, String?) }
func mediaID(_ value: String) throws -> String {
    guard value.count == 36, UUID(uuidString: value) != nil else { throw MediaError.invalid }; return value
}
enum MediaKind: String, Codable, Sendable {
    case photo = "PHOTO", video = "VIDEO", avatar = "AVATAR"
    var maximumBytes: Int64 { self == .video ? 50 * 1024 * 1024 : 10 * 1024 * 1024 }
    var contentTypes: Set<String> { self == .video ? ["video/mp4", "video/quicktime"] : ["image/jpeg", "image/png", "image/webp"] }
}
enum MediaStatus: String, Codable, Sendable { case reserved, uploading, processing, ready, deleting, deleted }
struct MediaReceipt: Codable, Sendable, Equatable {
    let assetId: String
    let status: MediaStatus
    func validated(assetID: String? = nil) throws -> Self {
        _ = try mediaID(assetId); guard assetID == nil || assetID == assetId else { throw MediaError.invalid }; return self
    }
}
struct MediaFile: Sendable {
    let url: URL
    let kind: MediaKind
    let contentType: String
    let byteLength: Int64
    init(url: URL, kind: MediaKind, contentType: String) throws {
        let values = try url.resourceValues(forKeys: [.fileSizeKey, .isRegularFileKey, .isSymbolicLinkKey])
        guard url.isFileURL, values.isRegularFile == true, values.isSymbolicLink != true,
              let size = values.fileSize, size > 0, size <= kind.maximumBytes, kind.contentTypes.contains(contentType) else { throw MediaError.invalid }
        self.url = url; self.kind = kind; self.contentType = contentType; self.byteLength = Int64(size)
    }
    func validate() throws {
        let current = try MediaFile(url: url, kind: kind, contentType: contentType)
        guard current.byteLength == byteLength else { throw MediaError.invalid }
    }
    func remove() { try? FileManager.default.removeItem(at: url) }
}
struct MediaIntent: Encodable, Sendable {
    let kind: MediaKind
    let contentType: String
    let byteLength: Int64
    let roomId: String?
    init(file: MediaFile, roomID: String?) throws {
        try file.validate()
        guard (file.kind == .avatar) == (roomID == nil) else { throw MediaError.invalid }
        if let roomID { _ = try mediaID(roomID) }
        kind = file.kind; contentType = file.contentType; byteLength = file.byteLength; roomId = roomID
    }
}
/** Existing outbox owns clientMessageId, intent, recipient and quote. This encodes only content. */
enum MediaContent: Encodable, Sendable {
    case attachments(MediaKind, [MediaReceipt])
    case sticker(String)
    private enum Keys: String, CodingKey { case type, assetIds, stickerId }
    func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: Keys.self)
        switch self {
        case .attachments(let kind, let receipts):
            let ids = try receipts.map { try $0.validated().assetId }
            guard kind != .avatar, !ids.isEmpty, ids.count <= (kind == .photo ? 4 : 1),
                  Set(ids).count == ids.count, receipts.allSatisfy({ $0.status == .ready }) else { throw MediaError.invalid }
            try c.encode(kind.rawValue, forKey: .type); try c.encode(ids, forKey: .assetIds)
        case .sticker(let id):
            try c.encode("STICKER", forKey: .type); try c.encode(mediaID(id), forKey: .stickerId)
        }
    }
}
enum MediaVariant: String, Codable, Sendable, Hashable { case image, video, poster }
enum MediaAccess: Sendable, Hashable {
    case preview(MediaVariant)
    case message(room: String, message: String, variant: MediaVariant)
    case avatar(room: String, actor: String)
    case sticker(room: String, sticker: String, message: String?)
    var variant: MediaVariant {
        switch self { case .preview(let v), .message(_, _, let v): v; case .avatar, .sticker: .image }
    }
    var roomID: String? {
        switch self { case .preview: nil; case .message(let r, _, _), .avatar(let r, _), .sticker(let r, _, _): r }
    }
    func body() throws -> Data {
        var values = ["variant": variant.rawValue]
        switch self {
        case .preview: break
        case .message(let room, let message, _): values["roomId"] = try mediaID(room); values["messageId"] = try mediaID(message)
        case .avatar(let room, let actor): values["roomId"] = try mediaID(room); values["actorId"] = try mediaID(actor)
        case .sticker(let room, let sticker, let message):
            values["roomId"] = try mediaID(room); values["stickerId"] = try mediaID(sticker)
            if let message { values["messageId"] = try mediaID(message) }
        }
        return try JSONEncoder().encode(values)
    }
}
struct MediaSticker: Decodable, Sendable, Identifiable { let id: String; let label: String; let assetId: String }
struct MediaStickerPage: Decodable, Sendable { let items: [MediaSticker]; let nextCursor: String? }
// Never Codable: signed URLs must not enter persistent stores, analytics, logs or SEND bodies.
struct MediaLease: Sendable, CustomStringConvertible {
    private let url: URL
    private let deadline: ContinuousClock.Instant
    let variant: MediaVariant
    var description: String { "MediaLease(redacted)" }
    init(url: URL, started: ContinuousClock.Instant, variant: MediaVariant) throws {
        guard url.scheme == "https", url.host != nil, url.user == nil, url.password == nil, url.fragment == nil else { throw MediaError.invalid }
        self.url = url; self.deadline = started.advanced(by: .seconds(55)); self.variant = variant
    }
    func renewalBudget(at now: ContinuousClock.Instant = .now) -> Duration { min(.seconds(5), now.duration(to: deadline)) }
    func needsRenewal(at now: ContinuousClock.Instant = .now) -> Bool { now >= deadline.advanced(by: .seconds(-20)) }
    func checkedURL(scope: any MediaScope) throws -> URL {
        try scope.check(); try Task.checkCancellation()
        guard ContinuousClock.now < deadline else { throw MediaError.expired }; return url
    }
}

struct MediaPresentationIdentity: Hashable { let scopeID: String; let assetID: String; let access: MediaAccess }

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/media/MediaContract.kt
package chat.rogi.rogichat.core.media

import java.io.File
import java.util.UUID
import kotlinx.serialization.json.*

/** Parent adapter captures the existing immutable account/ConversationScope, including epoch,
 * cache generation and original server M/A bindings. check must fail closed after invalidation. */
interface MediaScope {
    val presentationID: String
    val roomId: String?
    fun check()
}
fun mediaId(value: String): String { require(Regex("[0-9a-fA-F]{8}(-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}").matches(value)); UUID.fromString(value); return value }
enum class MediaKind(val maxBytes: Long, val types: Set<String>) {
    PHOTO(10L * 1024 * 1024, setOf("image/jpeg", "image/png", "image/webp")),
    VIDEO(50L * 1024 * 1024, setOf("video/mp4", "video/quicktime")),
    AVATAR(10L * 1024 * 1024, setOf("image/jpeg", "image/png", "image/webp"))
}
enum class MediaStatus { reserved, uploading, processing, ready, deleting, deleted }
data class MediaReceipt(val assetId: String, val status: MediaStatus) {
    init { mediaId(assetId) }
    companion object { fun decode(raw: String): MediaReceipt { val o = Json.parseToJsonElement(raw).jsonObject
        return MediaReceipt(o.getValue("assetId").jsonPrimitive.content, MediaStatus.valueOf(o.getValue("status").jsonPrimitive.content)) } }
}
class MediaFile(val file: File, val kind: MediaKind, val contentType: String) : AutoCloseable {
    val byteLength = file.length()
    init { require(file.isFile && byteLength in 1..kind.maxBytes && contentType in kind.types) }
    fun validate() { require(file.isFile && file.length() == byteLength) }
    override fun close() { file.delete() }
}
data class MediaIntent(val kind: MediaKind, val contentType: String, val byteLength: Long, val roomId: String?) {
    init { require(byteLength in 1..kind.maxBytes && contentType in kind.types)
        require((kind == MediaKind.AVATAR) == (roomId == null)); roomId?.let(::mediaId) }
    fun json() = buildJsonObject { put("kind", kind.name); put("contentType", contentType); put("byteLength", byteLength); roomId?.let { put("roomId", it) } }.toString()
}
/** Only content; existing room outbox owns the SEND envelope, intent and idempotency key. */
sealed interface MediaContent {
    fun json(): JsonObject
    class Attachment(val kind: MediaKind, receipts: List<MediaReceipt>) : MediaContent {
        val assetIds = receipts.map { it.assetId }
        init { require(kind != MediaKind.AVATAR && receipts.all { it.status == MediaStatus.ready })
            require(assetIds.size in 1..(if (kind == MediaKind.PHOTO) 4 else 1) && assetIds.distinct().size == assetIds.size) }
        override fun json() = buildJsonObject { put("type", kind.name); put("assetIds", JsonArray(assetIds.map(::JsonPrimitive))) }
    }
    class Sticker(val stickerId: String) : MediaContent {
        init { mediaId(stickerId) }
        override fun json() = buildJsonObject { put("type", "STICKER"); put("stickerId", stickerId) }
    }
}
enum class MediaVariant { image, video, poster }
sealed interface MediaAccess {
    val variant: MediaVariant
    fun json(): String
    data class Preview(override val variant: MediaVariant) : MediaAccess {
        override fun json() = buildJsonObject { put("variant", variant.name) }.toString()
    }
    data class Message(val roomId: String, val messageId: String, override val variant: MediaVariant) : MediaAccess {
        init { mediaId(roomId); mediaId(messageId) }
        override fun json() = buildJsonObject { put("roomId", roomId); put("messageId", messageId); put("variant", variant.name) }.toString()
    }
    data class Avatar(val roomId: String, val actorId: String) : MediaAccess {
        override val variant = MediaVariant.image
        init { mediaId(roomId); mediaId(actorId) }
        override fun json() = buildJsonObject { put("roomId", roomId); put("actorId", actorId); put("variant", variant.name) }.toString()
    }
    data class Sticker(val roomId: String, val stickerId: String, val messageId: String? = null) : MediaAccess {
        override val variant = MediaVariant.image
        init { mediaId(roomId); mediaId(stickerId); messageId?.let(::mediaId) }
        override fun json() = buildJsonObject { put("roomId", roomId); put("stickerId", stickerId); put("variant", variant.name); messageId?.let { put("messageId", it) } }.toString()
    }
}
data class MediaSticker(val id: String, val label: String, val assetId: String) { init { mediaId(id); mediaId(assetId) } }
data class MediaStickerPage(val items: List<MediaSticker>, val nextCursor: String?)
class MediaUnavailable : Exception("media_unavailable")
class MediaProcessingTimeout : Exception("media_processing")
class MediaFailure(val status: Int, val code: String?) : Exception("media_request_failed")
/** Deliberately not a data class: signed URLs must not appear in generated toString/logs. */
class MediaLease internal constructor(internal val url: String, private val expiresAtNanos: Long, val variant: MediaVariant) {
    fun renewalBudgetMillis(nowNanos: Long = System.nanoTime()): Long = minOf(5000, (expiresAtNanos - nowNanos) / 1_000_000).coerceAtLeast(0)
    fun needsRenewal(nowNanos: Long = System.nanoTime()): Boolean = nowNanos >= expiresAtNanos - 20_000_000_000L
    fun checkedURL(scope: MediaScope): String { scope.check(); check(System.nanoTime() < expiresAtNanos) { "media_access_expired" }; return url }
}

data class MediaPresentationIdentity(val scopeID: String, val assetId: String, val access: MediaAccess)
